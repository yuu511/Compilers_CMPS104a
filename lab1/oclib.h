// $Id: oclib.h,v 1.31 2019-04-11 12:36:19-07 - - $

#ifndef __OCLIB_H__
#define __OCLIB_H__

#define EXIT_SUCCESS 0
#define EXIT_FAILURE 1
#define EOF (-1)
#define TRUE 1
#define FALSE 0

#define assert(expr) \
        if (not (expr)) assert_fail (#expr, __FILE__, __LINE__); 

void assert_fail (string expr, string file, int line);

void putchr (int c);
void putint (int i);
void putstr (string s);

int getchr();
string getstr();
string getln();

void exit (int status);

#endif

