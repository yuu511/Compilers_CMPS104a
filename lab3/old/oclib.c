// $Id: oclib.c,v 1.3 2019-04-17 13:47:02-07 - - $

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define not !
#define nullptr 0
#define string char*

#include "oclib.h"

void fail (string expr, string file, int line) {
   fprintf (stderr, "%s:%d: assert (%s) failed\n", file, line, expr);
   abort();
}

void* xcalloc (int nelem, int size) {
   void* result = calloc (nelem, size);
   assert (result != nullptr);
   return result;
}

void putchr (int chr) { printf ("%c", chr); }
void putint (int num) { printf ("%d", num); }
void putstr (string str) { printf ("%s", str); }

int getchr() { return getchar(); }

string getstr (void) {
   static char buffer[0x1000];
   static char format[16];
   sprintf (format, "%%%zds", sizeof buffer - 1);
   int count = scanf (format, buffer);
   return count != 1 ? nullptr : strdup (buffer);
}

string getln (void) {
   static char buffer[0x1000];
   string result = fgets (buffer, sizeof buffer, stdin);
   return result == nullptr ? nullptr : strdup (result);
} 

