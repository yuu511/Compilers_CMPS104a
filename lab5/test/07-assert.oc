// $Id: 07-assert.oc,v 1.2 2019-04-16 13:14:01-07 - - $

#include "oclib.h"

int main() {
   assert ("nullptr" == nullptr);
   return FAILURE;
}

