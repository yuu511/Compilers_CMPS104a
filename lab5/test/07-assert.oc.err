not expression in loop may only be a single expression!
invalid condition for loop3ac: invalid statement passed :if
3ac: non-recognized expression parsed:';' 
 3ac: warning: expr parsed without assignment: ;:
3ac: invalid statement passed :;
NUMBER OF ERRORS REPORTED: 4
