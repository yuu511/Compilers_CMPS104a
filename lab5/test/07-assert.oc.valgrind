==24186== Memcheck, a memory error detector
==24186== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==24186== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==24186== Command: ../oc 07-assert.oc
==24186== 
==24186== 
==24186== HEAP SUMMARY:
==24186==     in use at exit: 0 bytes in 0 blocks
==24186==   total heap usage: 364 allocs, 364 frees, 138,050 bytes allocated
==24186== 
==24186== All heap blocks were freed -- no leaks are possible
==24186== 
==24186== For counts of detected and suppressed errors, rerun with: -v
==24186== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
