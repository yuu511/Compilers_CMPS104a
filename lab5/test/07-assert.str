string_set[   2]:    3835993668518112351 0x55e1d49d7fc8->"6"
                   12808232827189627082 0x55e1d49d7b78->"nullptr"
                    5344662657899890615 0x55e1d49d59b8->";"
string_set[   3]:   10159970873491820195 0x55e1d49d8148->"1"
string_set[   4]:    9201800968376702500 0x55e1d49d6b78->"getchr"
string_set[   7]:   13348923631108115385 0x55e1d49d6698->"putstr"
string_set[   8]:    9611428142580219387 0x55e1d49d50b8->"fail"
string_set[   9]:    5475249891658879256 0x55e1d49d7d58->""\"nullptr\" == nullptr""
                   18413295556186398326 0x55e1d49d6038->"putint"
string_set[  10]:   16579916199511856337 0x55e1d49d7178->"getln"
                   15816506932648033945 0x55e1d49d51b8->"("
string_set[  11]:   13686721134187162920 0x55e1d49d8088->"}"
string_set[  12]:   11868120863286193964 0x55e1d49d7478->"main"
                   12679825168387489523 0x55e1d49d61f8->"num"
string_set[  13]:   11166701219533395393 0x55e1d49d6e78->"getstr"
                    4706206694200543916 0x55e1d49d50f8->","
string_set[  14]:   16459639625557179573 0x55e1d49d7ed8->""07-assert.oc""
                    1583576014053361350 0x55e1d49d6858->"str"
                    7296690267450279874 0x55e1d49d5cd8->"chr"
                    4017852018407011111 0x55e1d49d55d8->"file"
string_set[  15]:   12654391512392035424 0x55e1d49d79f8->""nullptr""
                   12291392672316406109 0x55e1d49d5818->"line"
string_set[  16]:   12868445110721718657 0x55e1d49d58f8->")"
string_set[  19]:   14849783028996024063 0x55e1d49d78b8->"not"
string_set[  20]:    9317282556992397945 0x55e1d49d7778->"if"
                    9386170080689500517 0x55e1d49d5338->"expr"
string_set[  26]:   14635384904416252381 0x55e1d49d8108->"return"
string_set[  27]:   12597418120160561823 0x55e1d49d5758->"int"
string_set[  28]:    7211130878751977441 0x55e1d49d5b18->"putchr"
string_set[  29]:    3729804957429652673 0x55e1d49cecc8->"<<ROOT>>"
string_set[  31]:   14736352883882276958 0x55e1d49d7ab8->"=="
string_set[  32]:      21606623943142546 0x55e1d49d7638->"{"
string_set[  33]:    6142509188972423790 0x55e1d49d53f8->""
string_set[  35]:   11911796782241914524 0x55e1d49d3fe8->"void"
string_set[  36]:   13497578857207247412 0x55e1d49d5278->"string"
load_factor = 0.946
bucket_count = 37
max_bucket_size = 4
