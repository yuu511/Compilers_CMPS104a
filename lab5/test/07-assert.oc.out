CONDITIONAL IF STATEMENT SYMBOL TOK_NOT 
